filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,285 @@
+			sessionState.accessSession(func(session sarama.ConsumerGroupSession) {
+				if session == nil {
+					// The kafka connection is closed and / or is being rebalanced.
+					return
+				}
+				for _, event := range events {
+					if cm, ok := event.(*sarama.ConsumerMessage); ok {
+						session.MarkMessage(cm, "")
+					}
+				}
+			})
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "reading kafka input config")
+	}
+
+	saramaConfig, err := newSaramaConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing Sarama config")
+	}
+	consumerGroup, err :=
+		sarama.NewConsumerGroup(config.Hosts, config.GroupID, saramaConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing kafka consumer group")
+	}
+
+	input := &kafkaInput{
+		config:        config,
+		rawConfig:     cfg,
+		context:       inputContext,
+		outlet:        out,
+		consumerGroup: consumerGroup,
+		sessionState:  sessionState,
+		log:           logp.NewLogger("kafka input").With("hosts", config.Hosts),
+	}
+
+	return input, nil
+}
+
+// A helper to safely use the current sarama session for the duration of the
+// given callback. Used when ACKing messages outside the body of the main
+// sarama callbacks. The session parameter may be nil if there is no active
+// session.
+func (state *kafkaSessionState) accessSession(
+	fn func(session sarama.ConsumerGroupSession),
+) {
+	state.mutex.Lock()
+	state.waitGroup.Add(1)
+	session := state.session
+	state.mutex.Unlock()
+	defer state.waitGroup.Done()
+	fn(session)
+}
+
+// A helper to safely set the session field after waiting on any pending
+// operations.
+func (state *kafkaSessionState) setSession(sess sarama.ConsumerGroupSession) {
+	state.mutex.Lock()
+	// Once we claim the mutex we still wait for any pending ACKs to be
+	// sent. (These may well fail if the session is ending, but that's better
+	// than calling a stale pointer.)
+	state.waitGroup.Wait()
+	state.session = sess
+	state.mutex.Unlock()
+}
+
+// Run starts the input by scanning for incoming messages and errors.
+func (input *kafkaInput) Run() {
+	input.runOnce.Do(func() {
+		// Track errors
+		go func() {
+			for err := range input.consumerGroup.Errors() {
+				input.log.Errorw("Error reading from kafka", "error", err)
+			}
+		}()
+
+		go func() {
+			for {
+				handler := groupHandler{input: input}
+
+				// Sarama uses standard go contexts to control cancellation, so we need
+				// to wrap our input context channel in that interface.
+				err := input.consumerGroup.Consume(
+					doneChannelContext(input.context.Done), input.config.Topics, handler)
+				if err != nil {
+					input.log.Errorw("Kafka consume error", "error", err)
+				}
+
+				// If Consume returned because the context was cancelled, don't resume.
+				select {
+				case <-input.context.Done:
+					return
+				default:
+				}
+			}
+		}()
```

Comment:
I think you can simplify the loop to:

```
context := doneChannelContext(input.context.Done)
handler := groupHandler{input: input}
for context.Err() == nil {
  err := input.consumerGroup.Consume(context, ...)
  if err != nil {
    input.log.Errorw(...)
  }
}
```

Question: do we need some backoff between errors?

Reply:
Yes, this is much prettier. And yes, I think I will add backoff here along with the other pending run loop fixes.